src/CTA.WebForms2Blazor/Services/TaskManagerService.cs

```diff
@@ -0,0 +1,143 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace CTA.WebForms2Blazor.Services
+{
+    public class TaskManagerService
+    {
+        public enum TaskState { Active, Waiting }
+        public enum TaskStallTimeout { None = 0, Short = 100, Medium = 1000, Long = 10000 }
+
+        private IDictionary<int, ManagedTask> _managedTasks;
+        private bool _stalled;
+        private int _nextAvailableTaskId;
+
+        public TaskManagerService()
+        {
+            _managedTasks = new Dictionary<int, ManagedTask>();
+        }
+
+        public int RegisterNewTask()
+        {
+            var newManagedTask = new ManagedTask(_nextAvailableTaskId);
+            _nextAvailableTaskId += 1;
+
+            _managedTasks.Add(newManagedTask.TaskId, newManagedTask);
+            UpdateStallState();
+
+            return newManagedTask.TaskId;
+        }
+
+        public async Task<TResult> ManagedRun<TResult>(int taskId, Func<CancellationToken, Task<TResult>> func)
+        {
+            TResult result;
+            var managedTask = _managedTasks[taskId];
+            var token = managedTask.SetWaiting();
+            UpdateStallState();
+
+            try
+            {
+                result = await func(token);
+                managedTask.SetActive();
+            }
+            finally
+            {
+                UpdateStallState();
+            }
+
+            return result;
+        }
+
+        public void RetireTask(int taskId)
+        {
+            _managedTasks.Remove(taskId);
+            UpdateStallState();
+        }
+
+        private void UpdateStallState()
+        {
+            if (_managedTasks.Count > 0 && !_managedTasks.Values.Any(managedTask => managedTask.CurrentTaskState == TaskState.Active))
+            {
+                if (!_stalled)
+                {
+                    _stalled = true;
+                    TryResolveStall();
+                }
+            }
+            else
+            {
+                _stalled = false;
+            }
+        }
+
+        private async void TryResolveStall()
+        {
+            // Take the delay value as a constructor parameter?
+            // await Task.Delay((int)TaskStallTimeout.Short);
+            if (_stalled)
```

Comment:
Is this if statement necessary? Shouldn't you only be checking for stall outside the TryResolveStall function?

Reply:
The reason this if statement exists is because of the delay that has been commented out before it. If the delay were uncommented, then the if statement acts as a check that the status is still stalled after the timeout period has ended. I'll modify this to comment out the if statement, considering that it does nothing with the delay commented out